Add index mutation for GraphQL

eribeiro opened the pull request:
Fixes #543 - Expose ability to manage indexes in GraphQL

dougwettlaufer commented:
/gcbrun

eribeiro commented:
@dougwettlaufer Oops, sorry Doug, I forced pushed a commit to this branch soon after you run /gcbrun. Do you need to run it again?

dougwettlaufer commented on graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/ddl/CreateIndexFetcher.java:
```diff
@@ -0,0 +1,100 @@
+            .indexValues(indexValues)
+            .indexFull(indexFull)
+            .build();
+
```
Will also need to call authorizationService here, similar to

https://github.com/stargate/stargate/blob/master/graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/ddl/TableFetcher.java

dougwettlaufer commented on graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/ddl/DropIndexFetcher.java:
```diff
@@ -0,0 +1,55 @@
+      keyspaceName = parts.get(0);
+      indexName = parts.get(1);
+    }
+
```
Same comment regarding auth as above but will use scope DROP instead of CREATE.

dougwettlaufer reviewed (commented):
Small comments about authz

dougwettlaufer commented:
/gcbrun

eribeiro commented:
Ouch, looks like the building is failing but I don't have access to the full report. Maybe it's failing because of the changes to `QueryBuilderImpl`? 

I have identified some errors when it was building the CREATE INDEX command and an index name was supplied. Also when using custom indexes (CUSTOM).

PS: I have just submitted the changes as reviewed. :)

dougwettlaufer commented:
/gcbrun

dougwettlaufer commented on graphqlapi/src/main/java/io/stargate/graphql/schema/DdlSchemaBuilder.java:
```diff
@@ -453,6 +457,41 @@ private GraphQLFieldDefinition buildCreateTable() {
+            GraphQLArgument.newArgument().name("columnName").type(nonNull(Scalars.GraphQLString)))
+        .argument(GraphQLArgument.newArgument().name("indexName").type(Scalars.GraphQLString))
+        .argument(
+            GraphQLArgument.newArgument().name("customIndexClass").type(Scalars.GraphQLString))
```
I know under the hood in QueryBuilder it's `customIndexClass` but is that how we want to expose it? For example, an SAI index would be `StorageAttachedIndex` rather than a class like `org.apache.cassandra.index.sasi.SASIIndex`. 

Maybe something like 'indexType` and then add a description explaining it can be the class name or type of index?

dougwettlaufer commented on graphqlapi/src/main/java/io/stargate/graphql/schema/DdlSchemaBuilder.java:
```diff
@@ -453,6 +457,41 @@ private GraphQLFieldDefinition buildCreateTable() {
+        .argument(
+            GraphQLArgument.newArgument().name("customIndexClass").type(Scalars.GraphQLString))
+        .argument(GraphQLArgument.newArgument().name("ifNotExists").type(Scalars.GraphQLBoolean))
+        .argument(GraphQLArgument.newArgument().name("indexEntries").type(Scalars.GraphQLBoolean))
+        .argument(GraphQLArgument.newArgument().name("indexKeys").type(Scalars.GraphQLBoolean))
+        .argument(GraphQLArgument.newArgument().name("indexValues").type(Scalars.GraphQLBoolean))
+        .argument(GraphQLArgument.newArgument().name("indexFull").type(Scalars.GraphQLBoolean))
```
Might be good to surface these as descriptions

https://github.com/stargate/stargate/blob/master/persistence-api/src/main/java/io/stargate/db/query/builder/QueryBuilderImpl.java#L1049-L1062

dougwettlaufer commented:
> Ouch, looks like the building is failing but I don't have access to the full report. Maybe it's failing because of the changes to `QueryBuilderImpl`?
> 
> I have identified some errors when it was building the CREATE INDEX command and an index name was supplied. Also when using custom indexes (CUSTOM).
> 
> PS: I have just submitted the changes as reviewed. :)

Looks like the error is

```
[ERROR] Failed to execute goal com.coveo:fmt-maven-plugin:2.9:check (default) on project graphqlapi: Found 1 non-complying files, failing build -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
```

Quick fix is to run `mvn xml-format:xml-format fmt:format`

dougwettlaufer reviewed (commented):
Looks good overall, just a few small comments. Would like to get a second look from @olim7t when he gets back next week before merging.

eribeiro commented:
I have addressed the review comments. Please, let me know (and excuse me in advance) if I understood some comment incorrectly. :) I've also run the format command.

dougwettlaufer commented:
/gcbrun

olim7t commented on graphqlapi/src/main/java/io/stargate/graphql/schema/DdlSchemaBuilder.java:
```diff
@@ -453,6 +457,62 @@ private GraphQLFieldDefinition buildCreateTable() {
+            GraphQLArgument.newArgument()
+                .name("indexKeys")
+                .description("Index keys of a map")
+                .type(Scalars.GraphQLBoolean))
```
A GraphQL enum might be a bit more user-friendly:
```
enum IndexKind {
  KEYS
  VALUES
  ENTRIES
  FULL
}
```
For the default kind we can either keep the field nullable, or introduce a `SIMPLE` constant (that's how it's called in the Cassandra codebase).

eribeiro replied:
Please, let me know if I have addressed this review comment accordingly. I am still getting acquainted with GraphQL. :)

olim7t commented on graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/ddl/IndexFetcher.java:
```diff
@@ -0,0 +1,50 @@
+
+public abstract class IndexFetcher extends DdlQueryFetcher {
+
+  protected Scope scope;
```
Nit: if you're going to keep `getScope()`, you can turn this into a local variable.
Although I would personally remove `getScope()` and set it via the constructor instead.

eribeiro replied:
Done. 👍

olim7t commented on graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/ddl/CreateIndexFetcher.java:
```diff
@@ -0,0 +1,104 @@
+        throw new IllegalArgumentException("Missing field argument keyspaceName @ 'createIndex'");
+      }
+      keyspaceName = parts.get(0);
+      tableName = parts.get(1);
```
Why not, but if we do this here, we should revisit all the table and type fetchers for uniformity. I'm not sure if it's worth it, seems simple enough to pass the keyspace separately.

eribeiro replied:
I have removed the optional keyspaceName. 👍

olim7t commented on graphqlapi/src/main/java/io/stargate/graphql/schema/fetchers/ddl/CreateIndexFetcher.java:
```diff
@@ -0,0 +1,104 @@
+
+    String indexName =
+        dataFetchingEnvironment.getArgumentOrDefault(
+            "indexName", tableName + "_" + columnName + "_idx");
```
Don't generate a name if it's not provided, just omit the `builder.index()` call (passing `null` is also valid I think). It's optional, Cassandra will generate one.

Otherwise there is a subtle bug here: special characters are allowed in column names, but not in index names. For example:
```
mutation {
  createTable(
    keyspaceName: "test"
    tableName: "foo"
    partitionKeys: [{ name: "k", type: { basic: INT } }]
    values: [{ name: "the value", type: { basic: INT } }]
  )
  createIndex(keyspaceName: "test", tableName: "foo", columnName: "the value")
}
```
Produces the error:
```
Index name must not be empty, more than 2147483647 characters long, or contain non-alphanumeric-underscore characters (got \"foo_the value_idx\"
```

eribeiro replied:
@olim7t I have tried to address your review comments, but I am starting to learn GraphQL (and Stargate codebase), so please let me know if I misunderstood something. Particularly, the GraphQL enum declaration and usage. Thanks!

eribeiro commented:
@olim7t I have tried to address your review comments, but I am starting to learn GraphQL (and Stargate codebase), so please let me know if I misunderstood something. Particularly, the GraphQL enum declaration and usage. Thanks!

dougwettlaufer commented:
/gcbrun

olim7t commented:
Looks good, the GraphQL enum is what I had in mind 👍

eribeiro commented:
@olim7t Awesome! 😄 Btw, sincere question: this PR would suffice to PR #80, too?

PS: had to remove a last minute left over snippet.

dimas-b commented:
/gcbrun

dougwettlaufer commented:
FYI @polandll you can use this in your graphql examples now instead of dropping down to cql.
